pkg/agent/openflow/packetin.go

```diff
@@ -71,7 +77,7 @@ func (f *featureStartPacketIn) ListenPacketIn() {
 		case pktIn := <-f.subscribeCh:
 			// Ensure that the queue doesn't grow too big. This is NOT to provide an exact guarantee.
 			if f.packetInQueue.Len() < packetInQueueSize {
-				f.packetInQueue.Add(pktIn)
+				f.packetInQueue.AddRateLimited(pktIn)
```

Comment:
As far as I can tell from code, `AddRateLimited` adds a `waitFor` item to a channel of 1000 buckets if there is no available rate limter bucket, so it will be blocking when the channel is full. And it doesn't drop any items even it exceeds its rate limit. 
Maybe the length check against the queue can avoid appending more items when it exceeds the length of the queue. However the length of the ratelimitingQueue is the items that are ready to be processed (their ratelimiting waiting time have expired), not including the items in the `waitingForAddCh` channel. If the consumers process the ready items very quickly, it could happen that there are 1000 waiting items in the waitingForAddCh and 0 ready items in the queue, and a waiting item is poped from the channel and pushed to the queue every 10ms (if the rate limit is 100), and each incoming packet-in event will block here for 10ms when calling AddRateLimited and they will not be handled properly as it will only be handled after 10s (1000*10ms) when Traceflow request or connection already expire.

Reply:
We need to drop the packets here. Otherwise the backpressure will propagate to ofnet, and we will essentially block the loop that receives all messages from the switch (including bundle acknowledgements?): https://github.com/wenyingd/ofnet/blob/171b6795a2da8d488d38ae318ca3ce043481fc59/ofctrl/ofSwitch.go#L306

I am not a big fan of the current architecture: unless I am missing something, I think we have one channel / queue too many (`subscribeCh` and `packetInQueue`) and that everything could be done with a single channel.

My preference would be to change and simplify the code as follows:
- remove `packetInQueue` altogether
- make `subscribeCh` a buffered channel
- change `PacketRcvd` (ofctfl_bridge.go) to drop the packet if the subscribe channel is full (avoid backpressure to ofnet)
- when calling `SubscribePacketIn`, provide a rate limiter to control how fast packets can be dequeued by a consumer (this ensures low CPU usage of consumer)

maybe we can define a simple queue like this one:
```go
type PacketInQueue struct {
    rateLimiter *rate.Limiter
    packetsCh chan *ofctrl.PacketIn
}

func NewPacketInQueue(size int, r rate.Limit) *PacketInQueue {
    return &PacketInQueue{rateLimiter: rate.NewLimiter(r, 1), packetsCh: make(chan *ofctrl.PacketIn, size)}
}

func (q *PacketInQueue) AddOrDrop(packet *ofctrl.PacketIn) bool {
    select {
        case q.packetsCh <- packet:
            return true
        default:
            // channel is full
            return false
    }
}

func (q *PacketInQueue) GetRateLimited(stopCh <-chan struct{}) *ofctrl.PacketIn {
    when := q.rateLimiter.Reserve().Delay()
    t := time.NewTimer(when)
    defer t.Stop()
    for {
        select {
            case <- stopCh:
                return nil
            case <- t.C:
                break
       }
    }
    for {
        select {
            case <- stopCh:
                return nil
            case packet := <- q.packetsCh:
                return packet
       }
    }
}

// receiver side:
go func() {
    for {
        packet := q.GetRateLimited(stopCh)
        if packet == nil {
            return
        }
        // call all registered handlers
    }
}
```

Of course we should continue to have one instance of `PacketInQueue` per packet type so that Traceflow packets get their own queue.

Reply:
After my test (`hping3 --flood 10.10.1.114 -p 80 -2`), there is no significant difference in CPU usage whether we use this rateLimitQueue. I guess continuously calling `AddRateLimited` at a high rate also costs a lot of CPU? I think @antoninbas 's simplified code looks good and we can drop the packet over rate in this way. To make sure the Traceflow won't be choked or dropped by other packet-in messages, maybe make TraceFlow has its own queue would be better?

Reply:
As I wrote at the end, yes Traceflow should have its own queue

Reply:
My bad miss the end statement. 
Then let me try if this way can save the CPU usage.